server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardWriteLoadDistributionMetrics.java

```diff
@@ -0,0 +1,275 @@
+            return;
+        }
+
+        final var clusterState = clusterService.state();
+        final var shardWriteLoads = clusterInfo.getShardWriteLoads();
+        final var ingestNodeCount = (int) clusterState.nodes().stream().filter(this::isIndexingNode).count();
+        final var writeLoadDistributionValues = createPercentileArrays(percentiles.length, ingestNodeCount);
+        final var writeLoadPrioritisationThresholdValues = new ArrayList<DoubleWithAttributes>(ingestNodeCount);
+        final var shardCountsExceedingPrioritisationThresholdValues = new ArrayList<LongWithAttributes>(ingestNodeCount);
+        final var shardWriteLoadSumValues = new ArrayList<DoubleWithAttributes>(ingestNodeCount);
+        for (RoutingNode routingNode : clusterState.getRoutingNodes()) {
+            final var node = routingNode.node();
+            if (node == null || isIndexingNode(node) == false) {
+                continue;
+            }
+
+            double maxShardWriteLoad = Double.NEGATIVE_INFINITY;
+            double totalShardWriteLoad = 0.0;
+
+            shardWeightHistogram.reset();
+            try {
+                for (ShardRouting shardRouting : routingNode) {
+                    final double writeLoad = shardWriteLoads.getOrDefault(shardRouting.shardId(), 0.0);
+                    /*
+                     * Shard write-loads originate from an org.elasticsearch.common.metrics.ExponentiallyWeightedMovingRate,
+                     * they sometimes end up being calculated as very small (e.g. 3.3123178228374412E-21). These values
+                     * don't play nice with the HdrHistogram because it works best when there is a relatively small difference
+                     * in the scale of the values inserted into it.
+                     * They're also not very interesting, so we round them down to zero before adding them to the histogram.
+                     */
+                    shardWeightHistogram.recordValue(roundTinyValuesToZero(writeLoad));
+                    maxShardWriteLoad = Math.max(maxShardWriteLoad, writeLoad);
+                    totalShardWriteLoad += writeLoad;
+                }
+            } catch (ArrayIndexOutOfBoundsException e) {
+                // This shouldn't happen because our histogram should be auto-resizing, but just in case
+                final var message = "Failed to record shard write load distribution metrics for node "
+                    + node.getName()
+                    + " ("
+                    + e.getMessage()
+                    + ")";
+                assert false : message;
+                logger.error(message, e);
+                continue;
+            }
+
+            final var nodeAttrs = getAttributesForNode(node);
+
+            /*
+             * Only publish distribution and prioritization threshold metrics if the node contains at least one shard
+             */
+            if (Double.isFinite(maxShardWriteLoad)) {
```

Comment:
Optionally, I think this would work, too
```suggestion
            if (maxShardWriteLoad > Double.NEGATIVE_INFINITY) {
```
`Double.isFinite` is fine, I just had to look up / verify what it does.